File: src/wrappers/FFMpeg.java
package wrappers;

import dandere2x.Utilities.DandereUtils;

import java.io.File;
import java.io.IOException;
import java.io.PrintStream;

import static java.io.File.separator;

public class FFMpeg {

    /**
     * Extracts audio in the same regard as 'extract frames' would function.
     * For windows, ffmpeg is included by default, so one needs to manually
     * download it and set it's path, or just set the exectutables location as the 'ffMpeg' string
     *
     * @param workspace
     * @param timeFrame
     * @param fileName
     * @param audioLayer
     * @param duration
     */
    public static void extractAudio(PrintStream log, String ffmpegDir, String workspace, String timeFrame, String duration, String fileName, String audioLayer) {

        Process proc = null;
        Runtime run = Runtime.getRuntime();
        String command;

        if (DandereUtils.isLinux()) {
            log.println("extracting audio on linux..");
            command = "ffmpeg  -ss " + timeFrame + " -i " + fileName +
                    " -t " + duration + " -map " + audioLayer + " " + workspace + "audio.mp3";
        } else {
            log.println("extracting frames on windows...");
            command = "cmd.exe /C start " + ffmpegDir + " -ss " + timeFrame + " -i " + fileName +
                    " -t " + duration + " -map " + audioLayer + " " + workspace + "audio.mp3";
        }
        log.println("audio extraction command ..." + command);


        try {
            log.println("running command " + command);
            proc = run.exec(command);
        } catch (IOException e) {
            e.printStackTrace();
            log.println("fatal error: Could not execute extract audio! Running will continue");
            log.print(e.toString());
            System.out.println("fatal error: Could not execute extract audio! Running will continue");
        }

        synchronized (proc) {
            while (proc.isAlive()) {
                try {
                    log.println("proc is alive");
                    proc.wait(250);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
        synchronized (proc) {
            try {
                proc.wait(1000);
            } catch (InterruptedException e) {

            }
        }
    }


    public static void extractFrames(PrintStream log, String ffmpegDir, String workspace, String timeFrame,
                                     String fileName, String duration, int frameRate) {

        log.println("Extracting frames at " + workspace);
        new File(workspace + "inputs").mkdir();

        Process process = null;
        Runtime run = Runtime.getRuntime();

        String command;

        if (DandereUtils.isLinux()) {
            log.println("extracting frames on linux...");
            command = "ffmpeg  -ss " + timeFrame + " -i " + fileName + " -r " + frameRate + " -qscale:v 2 " +
                    " -t " + duration + " " + workspace + "inputs" + separator + "frame%01d.jpg";
        } else {
            log.println("extracting frames on windows..");
            command = "cmd.exe /C start " + ffmpegDir + " -ss " + timeFrame + " -i " + fileName + " -r "
                    + frameRate + " -qscale:v 2 " + " -t " + duration + " " + workspace + "inputs" + separator + "frame%01d.jpg";
        }

        log.println("command for frame extraction: " + command);


        try {
            log.println("running command " + command);
            process = run.exec(command);
        } catch (IOException e) {
            e.printStackTrace();
            log.println("fatal error: Could not execute extract frames!");
            log.print(e.toString());
            System.out.println("fatal error: Could not execute extract frames!");
        }

        try {
            process = run.exec(command);
        } catch (IOException e) {
            e.printStackTrace();
        }

        synchronized (process) {
            while (process.isAlive()) {
                try {
                    log.println("waiting for process to extract frames");
                    process.wait(250);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }

        log.println("extracted frames");
    }
}

File: src/wrappers/ffmpeg.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Name: Dandere2X FFMmpeg
Author: CardinalPanda
Date Created: March 22, 2019
Last Modified: April 2, 2019

Description: temp ffmpeg wrapper, terrible implementation fix later
"""
import subprocess

from context import Context


def extract_frames(context: Context):
    ffmpeg_dir = context.ffmpeg_dir
    time_frame = context.time_frame
    file_dir = context.file_dir
    frame_rate = context.frame_rate
    duration = context.duration
    input_frames_dir = context.input_frames_dir
    extension_type = context.extension_type


    exec = [ffmpeg_dir,
               "-ss",
               time_frame,
               "-i",
               file_dir,
               "-r",
               str(frame_rate),
               "-qscale:v",
               str(2),
               "-t",
               str(duration),
               "-vf",
               "noise=c1s=8:c0f=u",
               input_frames_dir + "frame%01d" + extension_type]


    print(exec)
    subprocess.run(exec)


def extract_audio(context: Context):
    ffmpeg_dir = context.ffmpeg_dir
    time_frame = context.time_frame
    file_dir = context.file_dir
    workspace = context.workspace
    duration = context.duration
    audio_layer = context.audio_layer
    audio_type = context.audio_type

    command = ffmpeg_dir + " -ss " + time_frame + " -i " + file_dir + \
              " -t " + duration + " -map " + audio_layer + " " + workspace + "audio" + audio_type

    exec = command.split(" ")
    print(exec)
    subprocess.run(exec)
